File: leetcode_oj/linklist/isPalindrome.py
'''
Given a singly linked list, determine if it is a palindrome.

Follow up:
Could you do it in O(n) time and O(1) space?

Hide Company Tags Amazon Facebook
Hide Tags Linked List Two Pointers
Hide Similar Problems (E) Palindrome Number (E) Valid Palindrome (E) Reverse Linked List
Difficulty: Easy
'''
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow = head
        fast = head
        rev = None
        while fast and fast.next:
            fast = fast.next.next
            tmp = rev 
            rev = slow
            slow = slow.next
            rev.next = tmp 
        if fast:
            slow = slow.next

        head = slow
        tmp = None
        tail = rev 
        ispal = True
        while slow and rev:
            if slow.val != rev.val:
                ispal = False
            slow = slow.next
            cur = rev 
            rev = rev.next
            cur.next = tmp 
            tmp = cur 
    
        #Joining two halfs
        if tail:
            tail.next = head
        return ispal

File: leetcode_oj/string/isPalindrome.cpp
'''
Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

For example,
"A man, a plan, a canal: Panama" is a palindrome.
"race a car" is not a palindrome.

Note:
Have you consider that the string might be empty? This is a good question to ask during an interview.

For the purpose of this problem, we define empty string as valid palindrome.

Hide Company Tags Microsoft Uber Facebook Zenefits
Hide Tags Two Pointers String
Hide Similar Problems (E) Palindrome Linked List
Difficulty: Easy
'''
class Solution {
public:
    bool isPalindrome(string s) {
        size_t len = s.length();
        int start = 0, end = len;
        while (start < end) {
            if(!isalnum(s[start])) {
                start++;
                continue;
            }   
            if(!isalnum(s[end])) {
                end--;
                continue;
            }   
            if(tolower(s[start]) != tolower(s[end])) {
                return false;
            }   
            start++; end--;
        }              
        return true;
    }
};
